class/practice_2.2/main.cpp

```diff
@@ -1,97 +1,119 @@
 
-bool readImage(const string& filename, vector<vector<Pixel>>& image, const string& outputFilename) {
-    ifstream inFile(filename);
-    ofstream outFile(outputFilename);
-    if (!inFile) {
-        cerr << "Unable to open input file." << endl;
-        return false;
-    }
+class Image {
+public:
+    Image() : image{} {} 
 
-    string line;
-    int row = 0;
-    while (getline(inFile, line)) {
-        stringstream ss(line);
-        vector<Pixel> pixelRow;
-        for (int col = 0; col < WIDTH; ++col) {
-            Pixel pixel;
-            char comma;
-            ss >> pixel.r >> comma >> pixel.g >> comma >> pixel.b;
-            if (comma != ',' || pixel.r < 0 || pixel.r > 255 || pixel.g < 0 || pixel.g > 255 || pixel.b < 0 || pixel.b > 255) {
-                cerr << "Invalid input format at row " << row + 1 << ", column " << col + 1 << "." << endl;
+    bool readImage(const string& filename, const string& outputFilename) {
+        ifstream inFile(filename);
+        ofstream outFile(outputFilename);
+        if (!inFile) {
+            cerr << "Unable to open input file." << endl;
+            return false;
+        }
+
+        for (int iRow = 0; iRow < HEIGHT; ++iRow) {
+            string line;
+            if (!getline(inFile, line)) {
+                cerr << "Unexpected end of file. Expected " << HEIGHT << " lines." << endl;
+                return false;
+            }
+
+            if (auto row = readRow(line)) {
+                image[iRow] = std::move(*row); 
+            } else {
+                cerr << "Error parsing row " << iRow + 1 << "." << endl;
                 return false;
             }
-            pixelRow.push_back(pixel);
         }
-        if (pixelRow.size() != WIDTH) {
-            cerr << "Input file should contain exactly " << WIDTH << " pixels per line." << endl;
+
+        if (inFile.fail()) {
+            cerr << "Error reading input file." << endl;
             return false;
         }
-        image.push_back(pixelRow);
-        ++row;
-    }
 
-    if (row != HEIGHT || image.size() != HEIGHT) {
-        cerr << "Input file should contain exactly " << HEIGHT << " lines." << endl;
-        if (outFile) {
-            outFile << "Input file should contain exactly " << HEIGHT << " lines.";
-        }
-        return false;
+        return true;
     }
 
-    return true;
-}
-
+    void writeImage(const string& filename) const {
+        ofstream outFile(filename);
+        if (!outFile) {
+            cerr << "Unable to open output file." << endl;
+            return;
+        }
 
-void writeImage(const string& filename, const vector<vector<Pixel>>& image) {
-    ofstream outFile(filename);
-    if (!outFile) {
-        cerr << "Unable to open output file." << endl;
-        return;
+        for (const auto& row : image) {
+            for (size_t col = 0; col < row.size(); ++col) {
+                outFile << row[col].r << ',' << row[col].g << ',' << row[col].b;
```

Comment:
std::println(outFile, "{},{},{}", row[col].r, row[col].g, row[col].b);
and you can use std::views::join_with(" ") to format the result string
As a result we can use something like that
```
for (const auto& row : image) {
   std::println( row
      | std::ranges::transform([](const auto& pixel) {
                                    return std::format("{},{},{} ", pixel.r, pixel.g, pixel.b);
                                })
      | std::views::join_with(" ")
      | std::ranges::to<std::string>()
   );
}
```